### ISSUE TITLE
Allow circular dependencies/references for resources in export var

### ISSUE THREAD

**Issue by rossunger** [godotengine/godot-proposals#7363]
### Describe the project you are working on

I'm working on a game with many characters where the data about their personal lives is stored in resources which need to be accessed from a variety of .tscn files. There are many times different characters need to reference each other e.g. I have an export var which is an array of family members, and all the family members have references to each other. 

### Describe the problem or limitation you are having in your project

Godot does not allow circular references in resources in export vars. 

I apologise if this is the wrong place to post about this, it seemed the most appropriate to me, as this isn't a bug (it seems intentional), and I would like to see if it's possible/desirable to change this. 

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I'm not understanding the reason why resources can't have circular dependencies. I've looked through the forums and the issues on github and I see several thread about making the editor prevent circular dependencies instead of crashing, but I haven't been able to find an explanation for why this happens in the first place. I'm hoping someone can point me to a discussion thread, or even some of the godot source code 

I understood the ref-counted issue, but it seems to have been resolved in  [#67714](https://github.com/godotengine/godot/pull/67714),

and I was looking at [68281](https://github.com/godotengine/godot/pull/68281)  and I'm not sure if this made it into 4.1.1, or whether this solves my problem... 

I would like to try to propose a solution to allow circular dependencies, but I'm hoping someone can help me understand the problem first. 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

I don't understand the problem well enough to propose a solution, but I wasn't sure where else to post about it because it's not really a bug... it seems intentional.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

there are some hacky work-arounds involving referencing the file-path instead of the resource itself, but this requires you to load the resource to access them, and it doesn't allow you to edit the resource in the editor's inspector

I could also not use resources at all, and instead create my own tool which allows me to edit data and then serialize it for saving, but this is basically recreating Resources just to add support for circular dependency

### Is there a reason why this should be core and not an add-on in the asset library?

Resources are a core component of the engine

**Comment by aqwalnut** [comment]
The problem appears to be mostly due to _export_ in particular, and not because circular dependencies are not supported. [68281](https://github.com/godotengine/godot/pull/68281) is still open, so it might be a work in progress?

A structure like this might be convenient, for example, if I want to serialize a linked list of Resources, or if I want to make and edit a directed graph of places to traverse. For simple problems like those, one can use an ID system, but if things get more complicated (for example if the vertices have types and restrictions on types of vertices they can be linked to because they trigger certain behaviors on the neighboring vertices), you may need a different ID system for every subtype or interface and that gets messy.

Here's an example script in C# that causes Godot (latest version) to freeze and crash upon clicking the checkbox for Generate and then trying to inspect the resources in the exampleVariants array (one needs to click on another node in the scene and come back to it to see the items in the array), as well as a test to see if things are working.

`
using Godot;
using Godot.Collections;
using System;

[Tool]
public partial class VariantTest : Node
{
    //Only exported structures are kept in scene
    [Export]
    private Array<ExampleVariant> exampleVariants = new Array<ExampleVariant>();

    [ExportGroup("Editor")]
    [Export]
    public bool Generate
    {
        set
        {
            if (value == true)
            {
                GD.Print("Tool: Generate microtest");
                exampleVariants.Add(new ExampleVariant(1));
                exampleVariants.Add(new ExampleVariant(2));
                exampleVariants.Add(new ExampleVariant(3));
                exampleVariants[0].A = exampleVariants[1];
                exampleVariants[0].B = exampleVariants[2];
                exampleVariants[1].B = exampleVariants[2];
                exampleVariants[2].B = exampleVariants[2];
                exampleVariants[2].A = new ExampleVariant(4);
            }
        }
        get
        {
            return false;
        }
    }
    [ExportGroup("Editor")]
    [Export]
    public bool Test
    {
        set
        {
            if (value == true)
            {
                GD.Print(exampleVariants[0].A.B.B.ID);
                exampleVariants[0].A.B.B.ID = 5;
                GD.Print(exampleVariants[0].B.ID);
                GD.Print("meow");
                GD.Print(exampleVariants[0].A.B.B.A.ID);
                exampleVariants[0].A.B.B.A.ID = 5;
                GD.Print(exampleVariants[0].B.A.ID);
            }
        }
        get
        {
            return false;
        }
    }
}

public partial class ExampleVariant : Resource
{
    public ExampleVariant(int ID)
    {
        this.ID = ID;
    }
    [Export] public int ID = 0;
    [Export] public ExampleVariant A;//the test works fine after removing [Export] here and in the next line.
    [Export] public ExampleVariant B;//
}
`

But without the [Export] attribute in A and B, the engine serializes the cyclic references fine and nothing crashes. Clicking the Test checkbox works as expected even after closing and reloading the scene.

The export/inspector problem seems like it would require design decisions... it might involve displaying the same resource multiple times in the nested mess (which run into issues with other copies of the same object not updating when you change one of them), or it might involve stopping upon detecting an cyclic reference with some priority scheme.

**Comment by AThousandShips** [comment]
What is the reason you want circular references? You shouldn't have them, you need to use weak references or you will have memory leaks and all sorts of problems, see [here](https://docs.godotengine.org/en/latest/tutorials/scripting/gdscript/gdscript_basics.html#memory-management)

**Comment by aqwalnut** [comment]
@AThousandShips That's helpful and makes my life somewhat easier, but it doesn't fix the crash in my example. The engine still accumulates errors and crashes if I replace the ExampleVariant class with weak references like so:

```
public partial class ExampleVariant : Resource
{
    public ExampleVariant(int ID)
    {
        this.ID = ID;
    }
    [Export] public int ID = 0;

    [Export]
    public ExampleVariant A
    {
        get
        {
            return (ExampleVariant) _A.GetRef();
        }
        set
        {
            _A = WeakRef(value);
        }
    }
    private WeakRef _A;

    [Export]
    public ExampleVariant B
    {
        get
        {
            return (ExampleVariant)_B.GetRef();
        }
        set
        {
            _B = WeakRef(value);
        }
    }
    private WeakRef _B;
}
```

**Comment by rossunger** [comment]
I would be very happy to use weakref if it serialized for saving and worked
as an export var type... but I don't think that's going to happen :p

I'm hoping someone can explain what happens in the saving process or the
export var process that makes circular references a problem. I was thinking
about trying to look it up in the c++ codebase but it's very overwhelming
and I'm not sure where to start.

I'm also not clear if Resources disallowing circular references is
intentional, or a temporary bandaid because a solution hasn't been made
yet. To me this feels like a very common use case for resources and I want
to understand what problems this causes. Godot seems to serialize packed
scenes with circular references excellently.

I'm considering rewriting all my resources as nodes and saving them as
packed scenes, but then I'm stuck using nodepaths, which means that they
all have to be in the same scene in order to work predictably.

**Comment by HeavensSword** [comment]
I believe I'm running into this issue as well. I'm attempting to port my Unity game into Godot, and I have a Scriptable Object for Jobs that I've translated into a Resource.

Each job can reference a Prereq Job and Advanced Job object of the same type. I can't link them in Godot though because it says "Recursion Detected" when attempting to set the reference. Meanwhile it works perfecting fine in Unity.

**Comment by Calandiel** [comment]
> I believe I'm running into this issue as well. I'm attempting to port my Unity game into Godot, and I have a Scriptable Object for Jobs that I've translated into a Resource.
> 
> Each job can reference a Prereq Job and Advanced Job object of the same type. I can't link them in Godot though because it says "Recursion Detected" when attempting to set the reference. Meanwhile it works perfecting fine in Unity.

You should use uIDs instead and get the typed resources with load or preload from the uID. Those allow circular references. It should be mentioned somewhere in the docs for people coming in from Unity but I don't think it is.

**Comment by rossunger** [comment]
> > I believe I'm running into this issue as well. I'm attempting to port my Unity game into Godot, and I have a Scriptable Object for Jobs that I've translated into a Resource.
> > Each job can reference a Prereq Job and Advanced Job object of the same type. I can't link them in Godot though because it says "Recursion Detected" when attempting to set the reference. Meanwhile it works perfecting fine in Unity.
> 
> You should use uIDs instead and get the typed resources with load or preload from the uID. Those allow circular references. It should be mentioned somewhere in the docs for people coming in from Unity but I don't think it is.

How would this work @export vars in the inspector?

**Comment by HeavensSword** [comment]
> > I believe I'm running into this issue as well. I'm attempting to port my Unity game into Godot, and I have a Scriptable Object for Jobs that I've translated into a Resource.
> > Each job can reference a Prereq Job and Advanced Job object of the same type. I can't link them in Godot though because it says "Recursion Detected" when attempting to set the reference. Meanwhile it works perfecting fine in Unity.
> 
> You should use uIDs instead and get the typed resources with load or preload from the uID. Those allow circular references. It should be mentioned somewhere in the docs for people coming in from Unity but I don't think it is.

Where can you find the uID of a resource?

**Comment by Go-Ba** [comment]
This is an issue that I ran into and is seriously making me consider leaving the engine and returning to unity until this is fixed. Making bi-directional/cyclical graphs with ScriptableObjects is a common use case for me for a variety of systems, and I have not found an alternative in godot which does not feel like a hacky workaround (such as using string references) which would be terrible to work with at scale.

For a quick example, if I wanted to make a dialogue system in unity and wanted a character to repeat themselves, I could very simply just have a `DialogueData` `ScriptableObject` reference itself as the next `DialogueData` to go to when the player chooses the `"please repeat that"` option. It seems that godot cannot match this functionality and it makes me concerned that I will not be able to build robust and extensible systems for larger games.

**Comment by Calandiel** [comment]
> This is an issue that I ran into and is seriously making me consider leaving the engine and returning to unity until this is fixed. Making bi-directional/cyclical graphs with ScriptableObjects is a common use case for me for a variety of systems, and I have not found an alternative in godot which does not feel like a hacky workaround (such as using string references) which would be terrible to work with at scale.
> 
> For a quick example, if I wanted to make a dialogue system in unity and wanted a character to repeat themselves, I could very simply just have a `DialogueData` `ScriptableObject` reference the first `DialogueData` as the next one to go to when the player chooses the `"please repeat that"` option. It seems that godot cannot match this functionality and it makes me concerned that I will not be able to build robust and extensible systems for larger games.

Use uids, they serialize to something similar to what unity stores internally for their references.
Alternatively, something like the Pandora plugin if you want a pretty ui for it.

I'm working on a first person open world rpg and was in a similar spot to you some time ago (see my posts earlier in this thread) but its been smooth sailing overall. Work with the tool, not against it.

**Comment by Go-Ba** [comment]
> Use uids, they serialize to something similar to what unity stores internally for their references.
Alternatively, something like the Pandora plugin if you want a pretty ui for it.

uIDs are cryptic strings, it seems impossible to see which resource that uID refers to from the editor which will make it extremely difficult to navigate your reference graph. I'll check out that addon, but I'd rather not have to rely on the development and upkeep of a random person's alpha addon for something as fundamental as being able to structure my data in a bidirectional graph

**Comment by Calandiel** [comment]
> > Use uids, they serialize to something similar to what unity stores internally for their references.
> > Alternatively, something like the Pandora plugin if you want a pretty ui for it.
> 
> uIDs are cryptic strings, it seems impossible to see which resource that uID refers to from the editor which will make it extremely difficult to navigate your reference graph. I'll check out that addon, but I'd rather not have to rely on the development and upkeep of a random person's alpha addon for something as fundamental as being able to structure my data in a bidirectional graph

If you want both bidirectional references *and* human readability, overwrite the way the field is displayed and display it directly. This is also what you had to do in Unity until version 5.2

See pages on customizing the editor with plugins and/or getters/setters.

A quick way to do it is to have a get/set for your uuid, then a second exported string under it (say, dialogue_uid and dialogue_name below it).
Then, the setter for uid can retrieve the name of the file (or an even more semantically meaningful name, if your data has one) and write it to dialogue_name, keeping it both human readable and allowing bidirectional references.

**Comment by dreadicon** [comment]
Going to add my 2cents here - while it is _generally_ not a best practice, circular references are a perfectly legitimate tool that are fine if used sparingly and correctly. Also, saying that a work-around exists doesn't mean a smoother implementation should be passed on. In this case, uID strings are prone to breaking silently, and requires a plugin to even be human readable. It may be an easy plugin, but it still requires the user to devote time and effort into learning the plugin system.

I ran into it with a network based map. The map has nodes the player can move between. Each node can connect to other nodes. It generally makes sense that each node would have a bi-directional reference to the nodes it connects to. Or lets take biomes; the nodes belong to certain biomes. When setting up a node's data as a resource, it should know what biome it's a part of, right? But biomes also need to know what nodes they can spawn. Thus, another circular reference. I could have made the biome keep track of the spawned nodes, but now the biome is stateful when that was entirely unnecessary if each node minded its own business.

Sure you can hack around it with things like brittle uIDs or asset path references. Or with a factory pattern. But what if you're not using a factory pattern because it doesn't suit the rest of what you're doing? The more high level architectures you mash together for particular cases, the less clear the architecture is and the less benefits you can reap. Also, because resources are not typically meant to be created and destroyed during runtime, is there really a problem with the objects being held onto in memory? At what point is it no longer the engine's design's job to police memory if that's such a concern in the edge case where resources COULD cause a memory leak?

Is this a must-have feature? Nope, not even close. But given how easy a solution is, and that it's a fairly common edge case, I still think this is a good feature request. +1 from me :)

**Comment by AThousandShips** [comment]
So how do you intend to handle the memory leaks of circular references? That applies regardless if it's stored or not, you simply *shouldn't* have circular references, they're not just "not a best pracitce" they're actually bad practice and *don't work*

If you mean using weak references that's a whole other thing, that's not circular references

**Comment by dreadicon** [comment]
> So how do you intend to handle the memory leaks of circular references? That applies regardless if it's stored or not, you simply _shouldn't_ have circular references, they're not just "not a best pracitce" they're actually bad practice and _don't work_
> 
> If you mean using weak references that's a whole other thing, that's not circular references

I've used circular references all the time in data holders; like in the example I gave, each mapNodeDefinition resource has a reference to a biomeDefinition resource. They are static data that isn't meant to change at any point during gameplay, nor should they go away. What's more, at least in .NET's case, it detects small circular references (a loop consisting of 2-3 objects) without much trouble during its mid-depth collection passes. 

I love weak references, but as you noted the engine doesn't understand them. There's no way to have a [WeakExport] option afaik (which would be a fantastic solution btw), and till someone approves the PR to support it for C# at least, there's not even Interface based support, which is Design By Contract 101 and a pillar of modern coding.

Really, it's not the end of the world, like I said. But there's perfectly legitimate use cases where the architectural work-around would introduce as much or more code complexity as prohibiting all forms of circular reference saves. A critical feature this is not, but a useful one it is, regardless of if it's restricted to Resources, a wrapper on uIDs with some added error-checking for broken references, or an implementation of WeakRef in some form - all of those would work, with the latter being probably the best solution.

**Comment by dreadicon** [comment]
> But in general this to me stems from actually misusing the system, resources _shouldn't_ link to the resource containing them, that's just not how resources should work, they _should_ be shared, and not care where they are in the relationship

How then would you suggest to store the game's read-only data for biomes which must know their map nodes so that map nodes can be generated, and a method for identifying the map node's biome is also required?

**Comment by AThousandShips** [comment]
Don't use a resource? Use some other data that isn't designed to be shared, resources are supposed to be shared and be location agnostic, that's their whole design and purpose

Or just use an identifier, I'm not sure about your specific game design, but the simple fact is that resources are *not* designed to be in this kind of direct relationship, you probably are using them for something they aren't designed for

Having:
* A resource that knows of the Node it's used by
* A resource that knows what resource it's used in

Are IMO likely indicators that you're using a resource for something it's not designed for, and you should instead use just an `Object` to handle your data, that saves you a lot of overhead anyway

**Comment by Go-Ba** [comment]
> and you should instead use just an `Object` to handle your data, that saves you a lot of overhead anyway

What is the method to instantiate an `Object` in the filesystem like a resource and edit it within the engine's inspector like a resource?

**Comment by dreadicon** [comment]
> Don't use a resource? Use some other data that isn't designed to be shared, resources are supposed to be shared and be location agnostic, that's their whole design and purpose
> 
> Or just use an identifier, I'm not sure about your specific game design, but the simple fact is that resources are _not_ designed to be in this kind of direct relationship, you probably are using them for something they aren't designed for
> 
> Having:
> 
> * A resource that knows of the Node it's used by
> * A resource that knows what resource it's used in
> 
> Are IMO likely indicators that you're using a resource for something it's not designed for, and you should instead use just an `Object` to handle your data, that saves you a lot of overhead anyway

Problem is POCOs and JSONs have no gui, no drag-drop, no visual wizard unless you build it, serialization and all. A resource not knowing what nodes it's used in is sensible; I view resources as your ROM; data configured before the game runs, or maybe deriving some results from a pre-set configuration, while Nodes hold state. 

But if Resources are stateless, why is it a problem to have circular references? I'm still a bit lost on that. I just can't picture a use for Resources which allows for unbound memory leaks; even if you clone one, it'll reference the original other half of the pair, which references the original that was cloned, not the new clone.

If Resources aren't for ROM data to be easily configured by designers and developers without any arbitrary limitations to that purpose, then what is? Tweaking stats and values, configuring sets and groups, this stuff is like 40% of what designers do in the editor. And for that matter, what use ARE resources useful for then from a designer perspective? Nothing about editor-time set circular dependencies with other resources makes them less share-able or location agnostic; why would it? If any resource can reference another resource, making that relationship reciprocal is only significant for garbage collection, which as I mentioned is only relevant to a slightly increased footprint, not an ongoing leak. Unless Resources are loaded and unloaded in the background when asked for by some process I'm not aware of?

**Comment by aqwalnut** [comment]
@dreadicon Can you link the PR for WeakExport? I want to follow it but can't find it.

**Comment by AThousandShips** [comment]
You get a memory leak with a circular reference, that's not an unbounded leak

Just that if A references B which references A they can never be freed as they are reference counted, see the page I linked above, it's basic memory management

**Comment by caimantilla** [comment]
My notifications are blowing up so just to chime in with an example from a shipped game of why this would be useful:
Cassette Beasts mostly uses resource references, with paths specifically being reserved for edge cases like a monster's reference to a loot table (which can potentially contain monsters). But one case which would cause a lot of pain made itself clear very quickly as I dug around: monsters store evolution data within themselves, as they should... using resource exports. If the developers ever want to add any kind of backwards evolution, this would need to be reworked, and mods might become incompatible.
Circular references might not be necessary from a code standpoint, but it's inevitable that they come up when designing a complex game as the only way to avoid convoluted solutions. I've personally had to swear off these kinds of references, instead toying with _get_property_list and static methods which return an enum-style hint for files in a directory, but this is still only appropriate for database-style use cases like a list of items, not level data like points in a map, and direct object refs would be much easier to manage... I still end up holding the ref in a member variable anyways for performance and convenience.

**Comment by AThousandShips** [comment]
To clarify a bit what I meant by how resources "should" be used, because I might have been a bit unclear, isn't that it's bad to use them in some ways, but that it's important to consider how they were designed to work

It's not bad in itself to have a two-way relationship with resources or objects, but resources weren't designed with that in mind or designed around that, so it's an unusual feature that you need to handle in your own way, because it doesn't fit the wider, main usage of resources

Now for *true* circular references, i.e. "`RefCounted` all the way down", I'd say there won't and shouldn't be any system for handling that, as they simply are not valid, you should always break the circle somewhere with a `WeakRef`

I'd say that adding support for serializing `WeakRef` references would be great, they'd simply be the same as the normal way but be deferred and handle just connecting it, possibly with some syntactic sugar to make it easier to use them without having the `get_ref` method

But for the "true" ones they simply aren't something that should be done, because they are bad code, because they don't work and leak memory

Note that a `RefCounted` referencing a non-`RefCounted`, like a `Node`, is *not* a circular reference and is perfectly valid

**Comment by dreadicon** [comment]
> @dreadicon Can you link the PR for WeakExport? I want to follow it but can't find it.

Sorry if my comment was a bit meandering, but I'm not aware of any implementation of a WeakExport or such - it was just an idea. The PR I was talking about is to add support for C# Interfaces, which are imho even more important than circular references designs.

> To clarify a bit what I meant by how resources "should" be used, because I might have been a bit unclear, isn't that it's bad to use them in some ways, but that it's important to consider how they were designed to work
> 
> It's not bad in itself to have a two-way relationship with resources or objects, but resources weren't designed with that in mind or designed around that, so it's an unusual feature that you need to handle in your own way, because it doesn't fit the wider, main usage of resources
> 
> Now for _true_ circular references, i.e. "`RefCounted` all the way down", I'd say there won't and shouldn't be any system for handling that, as they simply are not valid, you should always break the circle somewhere with a `WeakRef`
> 
> I'd say that adding support for serializing `WeakRef` references would be great, they'd simply be the same as the normal way but be deferred and handle just connecting it, possibly with some syntactic sugar to make it easier to use them without having the `get_ref` method
> 
> But for the "true" ones they simply aren't something that should be done, because they are bad code, because they don't work and leak memory
> 
> Note that a `RefCounted` referencing a non-`RefCounted`, like a `Node`, is _not_ a circular reference and is perfectly valid

I do get that, and I'll concede that even if a memory leak is or could be bound in scope it shouldn't really be supported in the core engine. But as myself and others have mentioned, Godot currently lacks a solution for circular references in data structures, and since Resources were designed with IMO a bit overly-broad use case in mind it may be appropriate to consider making a new type, either a sub-type of Resource or a new type all together that can handle circular references. Again, I agree WeakRef is ideal from a technical perspective in its simplicity and elegance, but I believe the implementation of that would be challenging for a variety of reasons (different languages handle (or dont handle) weak references differently, WeakRefs often require wrapping/unboxing every time they are accessed, etc). 

I do in fact know how memory leaks work; I previously worked on a custom in-house C# engine where memory leaks were a constant issue to track down and solve, and most recently built a custom event bus which uses weakref to eliminate all the boilerplate besides a parameterless ".subscribe()" call and has no way to leak memory (I love Godot's signals, but I want my architecture to be as platform agnostic as possible for reasons).  

So, all that to say I don't think we necessarily disagree on the practicalities of the design need or the practical hurdles in the end here - wrapping back to my original conclusion: Do I think this feature is the biggest bang for development buck in Godot? Nope. Would it be nice, practical, and useful in a number of cases? Yep. One thing I DONT want to see from my time in Unity is 'Sorry, that's the design we picked a long time ago so that feature is never happening' - imho Unity's unwillingness to moderately break backwards compatibility every 2-5 years in order to iterate on core elements was a major part of why it's so convoluted.

**Comment by AThousandShips** [comment]
Circular reference detection is error prone and expensive so I'd say it's better to just rely on users using references properly, over adding a lot of overhead to solve incorrect usage IMO

> Sorry, that's the design we picked a long time ago so that feature is never happening

And that's a valid desire, but it's not something to fear here, that's not what I'm saying :)

**Comment by Go-Ba** [comment]
It doesn't seem helpful to continually return to "it wasn't designed for that", "people just shouldn't do things that way" and "that's an incorrect usage of references" when it's been explained multiple times that:

1. Bidirectional graphs are a common and very elegant structure that allows you to do important things without much overhead
2. Resources and export fields are the only native tool for storing data about the game with proper integration in the engine's inspector
3. Other engines (i.e. Unity) already allow this pattern and it's very useful and intuitive.

If resources were specifically designed to have such limitations then they shouldn't be presented as the main and sole in-engine tool for storing data and a new one should be implemented. But I don't see what specific use a structure which only allows directed acyclic graphs has thus a discussion about changing the functionality of resources is instead taking place.

"users can just write their own solution" is not a valid answer to a lack of functionality in an engine. Especially when this is an error which a lot of beginners face because having two things reference each other is a normal and intuitive way to approach a lot of problems.

**Comment by AThousandShips** [comment]
SO I'll just leave a final comment here and then leave this discussion, and that is that this needs to be clarified a bit:
If we are talking about simply supporting actual circular references in resource exports, then there needs to be a dedicated proposal for adding some form of garbage collector or other management system to the engine core or GDScript *first*, because if you are wanting support for "Resource A has a reference to Resource B which has a reference to Resource A" then that's not something we will, or should, support with the current way the engine works, those circular references are *bad practice* and *do not work* and *will leak memory*. There's no nice way around it, those are not opinions but facts about how the engine works. That needs its own discussion, which is going to involve weighing the costs (because this kind of GC isn't free, nor is it perfect), and *then* we can discuss how that should work with resources.

Currently resources (at least text based ones) traverse down resource exports and exports the deepest resources first, this is because no resource should have a circular reference, so no need to have an elaborate system for this, it simply works because we assume people use references properly.

Now for the second option, which I fully support, namely to add support for serializing `WeakRef` properly, this wouldn't be too complicated IMO, and would work and wouldn't break things or be bad code.

With this you could then have either an automatic system exposing this property as if it was a non-weak reference, hiding that it is underneath, or you could use little helper exports to do it for you, like so:
```gdscript
@export var _my_export_weak : WeakRef

var my_export: MyResource:
    get():
        return _my_export_weak.get_ref()
    set(value):
        _my_export_weak = weakref(value)
```
And you'd just indirect the access to ensure it works correctly

That's all I have to say on this, hope you have a good day

**Comment by AThousandShips** [comment]
Did some experimentation with adding weak reference serialization and while I made good headway I've come to the view that it's less critical and that the relevant use cases are few, and that the cases when you really *do* need to have weak references to break chains they can be done with other means than built-in serialization support, if there's a real demand for the weak reference export support I can pick that work back up, but it'd involve breaking compatibility with the file formats and other issues that I feel aren't worth it and wouldn't be approved by reviewers

Instead I'd say that this should be handled by code, for example if you have a resource `A` that contains a resource `B` property and you need `B` to know `A` you should probably solve that by having `A` have a:
```
set(val): 
    my_val = val
    if is_instance_valid(my_val):
        my_val.my_parent = weakref(self)
```
To handle the relationship management, that would solve the need and would work just as well, without any complex changes to the resource writing and loading

Adding a system for weak references to serialization would, I believe, be possible, and not too difficult in my experimentation so far, but it would be potentially fragile, and I do believe this is a niche feature and can (at least in most cases) be worked around

The one area that is a bit weak currently is exporting arrays of resources, that is a case where you need some custom management to handle the relationship, as at least with the inspector you don't (currently) have specific set/get features to handle that

This can be handled in a few different ways, including custom array exports that use `_set/_get` to handle the actual data by paths, like `my_sub_resources/name`, etc., but other than that area I'd say that using setters to handle the parent relationship is a sufficient workaround for this, using `WeakRef` to manage it

If there's further interest in that part, or issues with using any of these workarounds, I'd be interested in looking back into this, but I don't currently think such an improvement would be approved, and I don't have the time to work more on it right now as it's slightly involved (though entirely possible)

**Comment by HeavensSword** [comment]
In the short-term I'd be happy with just a usability enhancement such as:

- The ability to specify that a string field is intended to be used as a UID, perhaps with an attribute/decorator.
- The ability to then drag an asset onto that exposed field and the editor to automatically fill it with that asset's UID.

Again, I'm not a Godot expert, but a Unity refugee; so if there is a quick way to do this myself or if it's possible to do this already in some manner, please share!

**Comment by aqwalnut** [comment]
Speaking of uid drag and drop.... is there a particular reason the "shows up in inspector" functionality of Export, and the "serialization and memory is managed this way" functionality of Export are tied together?

Just being able to separate these functionalities would make it easy to do a lot of things without needing to write a custom inspector... including the thing mentioned above.

**Comment by Natthapolmnc** [comment]
Hi, Idk if I try this correctly but I have made this code for the array of resource that support exporting.

```
@export var connectedStates:Array[Resource]:
	set(new_value):
		var val=[]
		for i in new_value:
			val.append(weakref(i))
		if is_instance_valid(connectedStates):
			connectedStates=val
	get:
		return connectedStates
```

it resulted in many lines of error like

```
E 0:00:00:0702   _parse_ext_resource: res://scripts/class/Movement/Move/PlayerMove/sprintState.tres:40 - Parse Error: [ext_resource] referenced non-existent resource at: res://scripts/class/Movement/Move/PlayerMove/walkState.tres
  <C++ Source>   scene/resources/resource_format_text.cpp:162 @ _parse_ext_resource()
E 0:00:00:0702   load: res://scripts/class/Movement/Move/PlayerMove/sprintState.tres:40 - Parse Error: [ext_resource] referenced non-existent resource at: res://scripts/class/Movement/Move/PlayerMove/walkState.tres
  <C++ Source>   scene/resources/resource_format_text.cpp:706 @ load()
E 0:00:00:0702   _load: Failed loading resource: res://scripts/class/Movement/Move/PlayerMove/sprintState.tres. Make sure resources have been imported by opening the project in the editor at least once.
  <C++ Error>    Condition "found" is true. Returning: Ref<Resource>()
  <C++ Source>   core/io/resource_loader.cpp:273 @ _load()
```

I don't know if I'm using the weakRef class correctly or I don't know the detail of is_instance_valid. Like I know it to check if the Object is in-memory but Idk when will the Object be deleted from the memory or instantiate in the memory. 

> Did some experimentation with adding weak reference serialization and while I made good headway I've come to the view that it's less critical and that the relevant use cases are few, and that the cases when you really _do_ need to have weak references to break chains they can be done with other means than built-in serialization support, if there's a real demand for the weak reference export support I can pick that work back up, but it'd involve breaking compatibility with the file formats and other issues that I feel aren't worth it and wouldn't be approved by reviewers
> 
> Instead I'd say that this should be handled by code, for example if you have a resource `A` that contains a resource `B` property and you need `B` to know `A` you should probably solve that by having `A` have a:
> 
> ```
> set(val): 
>     my_val = val
>     if is_instance_valid(my_val):
>         my_val.my_parent = weakref(self)
> ```
> 
> To handle the relationship management, that would solve the need and would work just as well, without any complex changes to the resource writing and loading
> 
> Adding a system for weak references to serialization would, I believe, be possible, and not too difficult in my experimentation so far, but it would be potentially fragile, and I do believe this is a niche feature and can (at least in most cases) be worked around
> 
> The one area that is a bit weak currently is exporting arrays of resources, that is a case where you need some custom management to handle the relationship, as at least with the inspector you don't (currently) have specific set/get features to handle that
> 
> This can be handled in a few different ways, including custom array exports that use `_set/_get` to handle the actual data by paths, like `my_sub_resources/name`, etc., but other than that area I'd say that using setters to handle the parent relationship is a sufficient workaround for this, using `WeakRef` to manage it
> 
> If there's further interest in that part, or issues with using any of these workarounds, I'd be interested in looking back into this, but I don't currently think such an improvement would be approved, and I don't have the time to work more on it right now as it's slightly involved (though entirely possible)
> 
> Tl;dr; There are workarounds for setting `WeakRef`s on assigning properties, proven to work as I literally used them to work with my attempts to write a system for serializing `WeakRef`

**Comment by AThousandShips** [comment]
That won't work, look at my code, you need a backing structure of `WeakRef`, further you can't export the non-weak one, as it'll still have circular references, I'd suggest:
```gdscript
var _connectedStatesWeak : Array[WeakRef]

var connectedStates:Array[Resource]:
	set(new_value):
		var val : Array[WeakRef]
		for i in new_value:
			val.append(weakref(i))
		_connectedStatesWeak = val
	get:
		var ret : Array[Resource]
		for i in _connectedStatesWeak:
			if is_instance_valid(i):
				ret.append(i.get_ref())
		return ret
```
But haven't tested

**Comment by Natthapolmnc** [comment]
I see so if I wanted it to be exported I have to export the array of string instead and used it to update the value of real array??

**Comment by robbertzzz** [comment]
I get that this is an issue in the editor due to how nested resources are displayed, but is there any reason why the file format itself doesn't allow for circular references? The .tscn format allows circular references since they're really just UID and/or NodePath strings, it seems odd that .tres sets the value to null. The editor not displaying a resource shouldn't stop the file format from storing a resource.

Note, this exact thing has a bug report at https://github.com/godotengine/godot/issues/89961, but I was referred here to discuss it.

**Comment by AThousandShips** [comment]
See above:
> Currently resources (at least text based ones) traverse down resource exports and exports the deepest resources first, this is because no resource should have a circular reference, so no need to have an elaborate system for this, it simply works because we assume people use references properly.

It would require reading and writing text resources in a non-linear way which is much more complicated, I've tried implementing it with weak resources but it was very fragile and relied on some hacks, so it's not something that's necessarily easy to implement 

The reason it isn't supported, and no one has written a non-linear storing of them in the text format is because they're not encouraged or supported in the system, so the serialisation can be made *much* cheaper and faster

If we either implement support for serialising weak references (what I strongly recommended) or add some way to handle circular references in the engine itself (which would be needed, as they're not currently supported *safely*) with a dedicated way to serialise them, though again this would be complex 

This was discussed some time ago in the contributor chat, when I was testing out a weak reference feature, but the support was low so I abandoned the idea as the core team didn't support the feature at the time

There are workarounds if you want either to use weak references or full reference cycles by storing paths or using initialization, some examples are available above in the thread

**Comment by robbertzzz** [comment]
Circular references just turn a tree structure into a graph structure, which aren't hard to serialise at all. The main difference is that you can no longer serialise from the bottom up as is the current approach as there's no clear "bottom". What you do have is a clear starting point, namely the resource being serialised, and from there on it's a matter of keeping track of which resources have already been serialised and which still need to be addressed. This is a very common problem with a straightforward solution, simply keep an open and closed set, loop over the open set, add unaddressed resources to the open set and move anything to closed when it has been addressed.

This approach is common in all kinds of graph traversals, but for an easy-to-find implementation of it I'd recommend the [A* algorithm](https://en.wikipedia.org/wiki/A*_search_algorithm#Pseudocode).

**Comment by AThousandShips** [comment]
If you have a workable solution you're welcome to open a PR for it, but I'd suggest using `WeakRef`s to break the cycles so it doesn't introduce circular references just by loading a resource, as I don't think that would be accepted, but the interest in this from the team has been low when previously raised

But again in my testing integrating this into the existing system without introducing weak points or more overhead is not trivial

**Comment by lazyvainglory** [comment]
I’ve been grappling with resource-related issues recently. While working on a dialogue system, I encountered frequent missing dependencies especially with my ‘conversation resources’ which references other 'conversation resources.' To address this, I transitioned to using .json files. Today, I attempted to create a state machine for my enemies using ‘state resources’ that can reference other ‘state resources,’ but it broke again. I’m genuinely puzzled. And here I thought Resources = Scriptable Object

**Comment by Limofeus** [comment]
I also think that allowing circular dependencies would be a huge benefit and would allow to better utilize resources, especially for state machines, having simple code:
```gdscript
class_name State
extends Resource

@export
var possible_next_states : Array[State] = []

func _on_enter() -> void:
	pass
	
func _on_exit() -> void:
	pass
	
func _update(_delta: float) -> void:
	pass

func _can_transition() -> bool:
	return false
```
You can only make linear state machines? Otherwise you'll get this kind of strange errors that dont even tell you that the reason why you cant open any of your resources or scenes utilizing them is a circular dependency.
![image](https://github.com/user-attachments/assets/401c35e5-a153-4569-bb0e-98e9964ae1bd)
 If not a way to allow circular dependencies (Maybe a way for user to mark variable as potential circular dependency) then a more reasonable error message would sure be appreciated.

**Comment by zadirion** [comment]
I'm here to add to the heaps of people that need some form of support for circular reference of Resources. If you have any decent amount of resources referencing each other, you're left unable to rename them without this proper support, unless you're willing to risk breaking references without knowing, if you rely on pure strings/integers as references

**Comment by rossunger** [comment]
Hi everyone. 

I get the sense that there's not going to be any traction on this question in the near future

I personally have developed my own system that uses sqlite for handling all data, and I used Godot to make a separate database editor that's specific to my use case, but could easily be modified for different data situations. This avoids the need for resources completely and is honestly a better system for managing data than resources. 

If people are interested in collaborating, I think it might be beneficial and more productive to create an open source add-on for Godot that deals with this in a user-friendly way.

**Comment by dxdesjardins** [comment]
Thought I would share a temporary non-ideal solution that's working for me. I'm using exported string ResourceUID's instead of the resources themselves. The exported packedSceneName is just to provide something human readable in the editor. You can right click any resource in the editor and copy/paste the ResourceUID into the export field.

Note: This is C#. If someone converts to GDScript, please share.

```C#
[Export] private string packedSceneName;
[Export] private string PackedSceneUID {
    get => packedSceneUID;
    set {
        packedSceneName = System.IO.Path.GetFileNameWithoutExtension(ResourceUid.GetIdPath(ResourceUid.TextToId(value)));
        packedSceneUID = value;
    }
}
private string packedSceneUID;

// Example of calling the resource in code (I use an extension method to make this less verbose)
PackedScene packedScene = GD.Load<PackedScene>(ResourceUid.GetIdPath(ResourceUid.TextToId(packedSceneUID)))
```

**Comment by AnidemDex** [comment]
> If someone converts to GDScript, please share.


Here you go:
```gdscript
@tool # must be tool, we are adding custom set/get
@export_file("*.tscn") var packed_scene_file_name:String
@export packed_scene_UID:String:
  set(value):
    if value.is_empty():
      packed_scene_file_name = ""
      _packed_scene_UID = ""
    else:
      _packed_scene_UID = value
      packed_scene_file_name = ResourceUID.get_id_path(ResourceUID.text_to_id(value))
    notify_property_list_changed()
   get:
      return _packed_scene_UID

func get_packed_scene() -> Resource:
  return load(ResourceUID.get_id_path(ResourceUID.text_to_id(_packed_scene_UID)))
```

Untested, but it technically should work :thinking: 


---
> (...) I'd say there won't and shouldn't be any system for handling that, as they simply are not valid (...)

Aw, this is bad, my entire event sheet relies on a kind of "what if node structure were resources" structure, where each resource can "contain" other resources, but those "contained" resources knows who holds them. No, I couldn't use resources because I use _a lot_ of those silly objects, and no, I didn't wanted to deal with memory management that would imply using plain `Object`

**Comment by ikisoft** [comment]
This is a huge downside for managing any type of data with resources. You'll almost always need to have circular references, works with other engines, and from a data standpoint I don't understand why it wouldn't be possible?

**Comment by AnidemDex** [comment]
Because that was not the thing resources were designed for if I don't missunderstand this issue thread. 

The solution would be making weakref serializable or making a new type of serializable object afaik

**Comment by sergioabreu-g** [comment]
I don't understand how an issue so basic and with so much support has been ignored for more than a year and raises no interest from the team.

Circular references are a basic feature of any decent programming language, they're not something `you shouldn't want` and that `will break your project in all sort of unspeakable ways`.

**Comment by azulika** [comment]
Still no progress on this??

**Comment by Calinou** [comment]
> Still no progress on this??

To my knowledge, nobody is currently working on implementing this.

**Comment by rossunger** [comment]
> > Still no progress on this??
> 
> To my knowledge, nobody is currently working on implementing this.

Im using sqlite to use databases instead of resources. It's very robust and much faster. 

Sqlite is a incredibly tiny library, but gdextension makes it many times bigger. @Calinou is there any chance the Godot team would consider making sqlite part of core and just bind the "query" function for Gdscript? 

I think there's a fairly steep learning curve for how to use SQL well, and how to build table schemas... But I think there's a way to design it with InspectorPlugins so that it feels seamless with the rest of the editor ... Maybe it just needs a good sensible default database for common use things like inventory, NPCs, etc with premade SQL queries and examples...

**Comment by AThousandShips** [comment]
What would replacing `Resource` with sqlite solve? This doesn't solve these problems, how would you handle memory management? Why is a database the way to handle non-database information?

**Comment by rossunger** [comment]
Databases have no problems with circular dependencies. Everything has an ID and and references other things by ID (which has a similar effect to weak references). With an inspector plugin we can make an interface for selecting and assigning data that completely bypasses the need for resources (which weren't really designed for this use case anyway?) 

With SQL there is no memory management... You just run queries any time you need to access data, which only becomes inefficient when reaching hundreds of thousands or millions of rows.

You wouldn't necessarily do this on every _process frame, but every second or two, no problem. It's blazing fast. 

The data is completely separated from implementation, and it's very easy to retrieve the data and "rotate" it how you need.

E.g. if I have 10 characters and they belong to 3 families, with resources it's impossible to structure properly because if family is a resource, you want to know which character belong to that family, but you also want your characters to know which families they belong to. This causes circular references. 

With database, the database is the central authority, and holds both characters and families and everything you need to know about them

You could definitely achieve something like this on gdscript with a singleton type of class, but it's MUCH slower, and very messy. You'd basically be reinventing databases if you tried to do it with gdscript, and your miss out on the incredibly efficiency of SQL. 

Also, SQL has good data validation to prevent you from making mistakes... Foreign key and primary key rules can ensure that your references are valid

**Comment by AThousandShips** [comment]
Resources don't work that way though, they're just objects, not entries in a database, this would add a whole layer of data fetching that's completely unnecessary, it would add a huge performance bottleneck

It might not become really inefficient with a lot of data, but it's never going to be as efficient as simple data stored in memory without abstraction

This also assumes relationships between `Resource`s that they aren't designed for

**Comment by sergioabreu-g** [comment]
Resources are nothing more than objects that are well integrated into the engine. Even though objects can hold data, they are not data themselves, trying to replace one with the other just doesn't make any sense. I agree that circular dependencies are essential in making resources useful for real games, which generally have complex and intertwined object relationships, but databases are definitely not the way to do that.

The type of data and relationships databases store is rarely useful in videogames. Those who want/need to use a database can integrate it themselves through GDExtension, I don't think that should be a core feature in any game engine.

**Comment by rossunger** [comment]
Thanks for taking the time to share your opinion/experience. 

I have integrated database with gdextension, so no complaints from me! It's be wildly effective and I think it could be useful for many games if people considered it an option. 

From my understanding, which may be inaccurate, SQL is incredibly efficiency compared to anything someone could implement in gdscript, and even in cpp, unless you have specialized knowledge about how to work with data, I think you'd have a hard time improving on SQL. The memory management will be more efficient WITHOUT databases, but searching/modifying the data is likely to be faster with SQL. 

I want to clarify that the benefit isn't in replacing objects or resources, but in leveraging the search capabilities. Anything that involves "simulations" in games becomes a nightmare to deal with because accessing the data requires iterating over arrays and dictionaries, and often the data ends up decentralised, which makes it hard to work with. One solution is to have a singleton type of object that stores all the data, but my argument is that that is just a database without the convenience or efficiency of SQL 

Also, I think something you're misunderstanding is that we KNOW that we're trying to use resources for something that they're not designed for. That's the whole point. We need a system for working with data that is convenient and simple enough for people without experience in designing data structures to use. 

That's why I decided to move away from resources and towards databases. We need ways to store data that isn't attached to nodes (because nodes have massive overhead) and using objects requires memory management, which is not ideal, and then your have to implement serialization for saving/loading, which = resources

Here are some example use cases that I think database would be better than resources:
- dialog systems, with different languages and audio and images
- advanved branching narratives
- quest systems
- inventory/shops
- npc "ai" and relationship related data 
- metrics and analytics for ability to respond to play style / player history

I think it's important to acknowledge that many Godot users are going to lack the skillset to architect data structures well, me include. I spent a long time confused about how to deal with data in Godot. There's not really a clear built in system. Databases became that for me. If there's another way, I would love to hear about it!

**Comment by sergioabreu-g** [comment]
If you are not saying resources should be replaced with a database, then I think this discussion is out of point. Even if a database were to be integrated into Godot, we'd still have the circular dependency issue with resources, which I think cannot be ignored because it's fundamental for building object relationships properly. I'm going to answer your arguments anyway just for fun xD, but if you want to go ahead with proposing a database integration in Godot I think you should do it in another proposal.

I agree with you that databases are efficient in handling large sets of structured data, I don't think anyone will argue with that. The problem is that they're not the right tool for the job. Games don't *generally* require such amounts of data, especially not indie games, which is what's going to be built with Godot 99.9% of the time. Database querying only becomes interesting in terms of speed once you're handling tens of thousands of objects and you need parallel access and data consistency, which is not something that happens usually in games. You can have good enough performance iterating over lists of thousands of objects in GDScript (not to mention C++...). You can even do it every frame if you want, GPU performance is always going to be what determines frame time unless you do something crazy or you're loading stuff, which is slow due to disk access and memory allocation. Also, if you say that most beginners don't have the skillset to architect data structures, why do you assume they're going to have the skillset to leverage an integrated database? I just don't see this makes much sense. Databases can be very useful for certain projects, which can use a plugin like you do, but they're not a good general solution for managing data in games.

**Comment by rossunger** [comment]
Thank you for answering! 

I think it would make sense to open another proposal. Circular references is a square hole solution for a round problem. On the surface, resources SEEM like the right tool for storing data. They have so many of the features needed, including serialization, memory management, editor integration with the inspector.

It feels like the better solution would be to make a separate "data management" system from Godot which does some magic internally for handling memory management, searching/querying, and saving data. It could theoretically use resource behind the scenes.

I think the reason we asked for circular dependencies for resources is because that seemes like a much more achievable ask than to ask for a whole new system to be accepted.

I agree with you that databases are not the BEST solution, but they handle 80% of the task, masking the implementation 80% easier, meaning for me they were the most ACHIEVABLE solution. It was hexagon hole for a round problem.

I would also argue that in games that do any sort of simulation like City builders, or where you're managing large quantities of NPCs there could easily be tens of thousands of row of data that needs to be used in parallel and maintain integrity. I think these are types of games that indie developers would make more of if they had good tools for data management.

While I don't think users would succeed with raw databases, I think there's a way to create an editor interface with templates and prebuilt queries that would set users up for success. And if it can become integrated with the inspector then it could become very user friendly 

I think I'll open a new pr. Do you know of anyone else having made a similar or related proposal so I'm not doubling up and wasting moderators time?

**Comment by Calinou** [comment]
> It feels like the better solution would be to make a separate "data management" system from Godot which does some magic internally for handling memory management, searching/querying, and saving data. It could theoretically use resource behind the scenes.

There are several related proposals out there currently, like https://github.com/godotengine/godot-proposals/issues/13.

**Comment by MartinMajewski** [comment]
I just started an application project in Godot (not a game but an app) and played with a reactive UI pattern that makes data changes visible to all nodes connected to a change signal. However, because Resources can have Child-Resources and the parents shall also emit a change signal when the child is changed, the child needs to know who the parent is to call the parent's change signal.

Now I got really frustrated! It would be so easy to save the entire state of the app with the ResourceSaver if circular references would be supported. 

But nooooo!

At least I am not the only one who is annoyed by that limitation.

**Comment by Tarqrafi** [comment]
_The following text has been translated, so the wording may not be entirely accurate._

### Allowing circular references is necessary and essential

#### Implementing circular references is a common feature of file management.
For conventional file management systems, cyclic references between files and even self-references are common occurrences.
For example:
- Soft and Hard Links in Operating Systems
- Internal links in Markdown
- Sheet references in Excel
- ScriptableObject references in Unity

In general operations, you can reference by simply dragging and dropping or holding the Alt key while dragging. For more complex cases, you might need to write a bit of formula or take a slight detour, but it won't be too complicated overall.

#### Handling circular references is also a necessary feature of data processing tools.
For a qualified data processing tool, it should be able to handle conventional data structures. However, if it cannot handle circular references, it means it cannot process structures more complex than a tree, such as doubly linked lists or graphs. 

In other words, even if we set aside some complex situations, a simple circular patrol route cannot be implemented directly using just files.

#### Implementing circular references is a basic feature of game development tools
Two-way passages in games are very common, such as stairs or doors.
A->B and B->A
There's no need to use Unity or UE as examples.
Even in RPG Maker, a two-way door or staircase isn’t very complicated.

**Comment by dodgyville** [comment]
I created resources with many-to-many relationships and just hit this circular saving issue. It's unfortunate that the many-to-many set up with resources works so well in the engine (since they are just arrays of pointers) but the structure is destroyed when using ResourceSave!

**Comment by zarlin** [comment]
I ran into the circular dependency issue as well, and found a workaround. This stores only the resource UID, and uses `_ValidateProperty` to show the resource field in the editor without serializing it to file. Note that I use PackedScene here, but it should work for most resources.

```cs
[Export]
private PackedScene TargetScene
{
	get
	{
		if (string.IsNullOrWhiteSpace(targetSceneUID))
		{
			if (!Engine.IsEditorHint())
				GD.PrintErr("No valid target scene set!");
			return null;
		}

		string path = ResourceUid.UidToPath(targetSceneUID);
		return ResourceLoader.Load<PackedScene>(path);
	}
	set
	{
		//set only in editor when valid scene gets assigned
		if (!Engine.IsEditorHint())
			return;

		targetSceneUID = value != null ? ResourceUid.PathToUid(value.ResourcePath) : null;
		NotifyPropertyListChanged();
	}
}

/// <summary>
/// UID of the target scene, used for serializing without cyclic references
/// </summary>
[Export]
private string targetSceneUID;

public override void _ValidateProperty(global::Godot.Collections.Dictionary property)
{
	//flag targetscnee property as editor-only
	if (property["name"].AsStringName() == PropertyName.TargetScene)
		property["usage"] = (int)PropertyUsageFlags.Editor;

	//flag UID as readonly
	if (property["name"].AsStringName() == PropertyName.targetSceneUID)
		property["usage"] = (int)(PropertyUsageFlags.ReadOnly | property["usage"].As<PropertyUsageFlags>());
}
```

While writing this I also though that this could be a way to solve it (PropertyHint.ResourceUid doesn't exist yet), which would show a resource reference in the inspector, but serializes only the UID string.

```cs
[Export(PropertyHint.ResourceUid)]
private string resourceUID;
```
